rfcs/20230704-dynamism-101.md

```diff
@@ -0,0 +1,825 @@
+* **Dimension sizes** (or **sizes**) to refer to sizes of dimensions,
+  e.g. "`tensor<16x?xf32>` has the following dimension sizes: 16 and unknown".
+* Unqualified "dimension" will not be used at all.
+
+By the virtue of being bootstrapped from MHLO, the StableHLO dialect already
+has support for **dynamism within types**:
+
+* **Unbounded dynamism**: In a tensor type, some or all dimension sizes may be
+  unknown (aka "dynamic"). In MLIR syntax, these dimension sizes are expressed
+  via question marks.
+* **Bounded dynamism**: The same but some of dynamic dimensions sizes may have
+  known upper bounds. In MLIR syntax, these dimension sizes are expressed via
+  question marks, and bounds are expressed via `#stablehlo.bounds`.
+* **Unranked dynamism**: In a tensor type, rank may be unknown. In MLIR
+  syntax, this fact is expressed via asterisks.
+
+```mlir
+// Static shapes:
+// All dimension sizes are known.
+%0 = stablehlo.add %arg0, %arg1 : tensor<16x16xf32>
+
+// Unbounded dynamism:
+// First dimension size is unknown (?).
+// Second dimension size is known (16).
+%1 = stablehlo.add %arg0, %arg1 : tensor<?x16xf32>
+
+// Bounded dynamism:
+// First dimension size is unknown (?), but its bound is known (16).
+// Second dimension size is known (16), so its bound is N/A (?).
+%2 = stablehlo.add %arg0, %arg1 : tensor<?x16xf32, #stablehlo.bounds<16, ?>>
+
+// Unranked dynamism:
+// The rank is unknown (*).
+%3 = stablehlo.add %arg0, %arg1 : tensor<*xf32>
+```
+
+Similarly, the StableHLO dialect already has support for **dynamism within
+operations**, with some ops such as PadOp having dynamic counterparts such as
+DynamicPadOp. For example:
+
+```mlir
+// "vanilla" PadOp:
+// low, high and interior paddings are implemented via MLIR attributes,
+// i.e. they are static.
+%0 = stablehlo.pad %arg0, %arg1, low = [1, 1], high = [1, 1], interior = [0, 0]
+  : (tensor<16x16xf32>, tensor<f32>) -> tensor<20x20xf32>
+
+// DynamicPadOp:
+// low, high and interior paddings are implemented via MLIR operands,
+// i.e. they are dynamic.
+%1 = stablehlo.dynamic_pad %arg0, %arg1, %arg2, %arg3, %arg4 :
+  : (tensor<?x?xf32>, tensor<f32>, tensor<2xindex>, tensor<2xindex>, tensor<2xindex>) -> tensor<?x?xf32>
+```
+
+Finally, dynamism (both within types and within operations) creates new
+opportunities for errors. For example, the semantics of elementwise operations
+like `add` are only defined when inputs and outputs have the same shape.
+For the static shape case, this can be checked **at compile time** (i.e. before
+the execution of the StableHLO program). However, when dynamism is involved,
+this can only be checked **at run time**.
+
+If some operation doesn't make sense at run time because some unknown ranks
+and/or unknown dimension sizes turned out to be incompatible with the operation
+and/or with each other, an error condition called **"shape mismatch"** occurs.
+In order to guard against shape mismatches, StableHLO programs may employ
+**shape checks**.
```

Comment:
I may be overindexing from my JavaScript days where they basically do the same thing.

This would require runtime support to bail out when a shape mismatch occurs. Are these shape checks inserted by the runtime as its own internal IR or does StableHLO have an op to represent shape checks? Or should front ends insert:

if (!Shape(op, expected)) return error() type logic?

Reply:
Nevermind, resolved via reading P2.